### ISSUE TITLE
arm64 cpu identification is not portable to BSDs

### ISSUE THREAD

**Issue by j-bm** [flame/blis#795]
The bli_cpuid.c code for ARM64 is fairly complex, relies on linux-isms and therefore is not portable to *BSD systems.

There is some #ifdef for Apple that seems redundant, though I must say I am not able to test this.

**Comment by devinamatthews** [comment]
Yes this is a known problem. Do you happen to know what the equivalent mechanism to `HWCAP` is on BSD?

**Comment by j-bm** [comment]
I'm slowly becoming more knowledgeable about aarch64 machines...

It seems Linux user-mode runs at EL1 allowing access to CPU feature registers like MIDR or ID_AA64PFR0_EL1 to succeed.

OpenBSD (at least) runs at EL0 therefore these instructions fail with SIGILL (illegal instruction).

I see two paths forward: grep the dmesg output for indicators; OpenBSD for now does not publish the full CPU feature set in the dmesg output for aarch64 (it does so for most other architectures.)

Or run test code, trap SIGILL and revert to "generic" if needed.

Both look about as hacky as can be imagined.

(I don't know what Apple machines do.)  I'm testing on Raspberry Pi 4B.

**Comment by j-bm** [comment]
OpenBSD is adding a FreeBSD feature elf_aux_info(3) which is much better constrained over the equivalent Linux getauxinfo(3).  It focuses on CPU features.

But there are so many arm64 simd (asimd) variations, and SVE as well.  Not all are supported(*) and for now, for example, SVE is not supported on OpenBSD.

(*) Support means that not only the hardware has the feature, but the operating system software supports it, for example in context switches the OS must save/restore vector registers.

**Comment by volyrique** [comment]
> It seems Linux user-mode runs at EL1 allowing access to CPU feature registers like MIDR or ID_AA64PFR0_EL1 to succeed.

No, that would be a complete violation of the OS security model. Rather, it is a case of the trap-and-emulate approach - refer to the documentation [here](https://docs.kernel.org/arch/arm64/cpu-feature-registers.html). It is also a potentially portable, albeit slow, mechanism (given that it is a mandatory trap to EL1), as long as other operating systems choose to implement it as well.
